Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,262 @@
+# HVPA-EP-0001: HVPA Redesign
+
+## Summary
+This proposal outlines a plan to enable HVPA to do more efficient scaling horizontally and vertically.
+
+## Motivation
+Currently, HVPA scales by applying user provided weights to HPA and VPA recommendations. We need to provide a better control over scaling of the applications to the users.
+
+### Goals
+
+1. HVPA should be able to scale an application with finer-grain control and predictability over horizontal and vertical scaling which can be specified by the users who are deploying/managing the application.
+
+## Approach
+
+Users will provide buckets of resources each of which would include minimum and maximum CPU, memory and replicas respectively. A configuration is said to belong to a bucket if total resources allocated to all its replcas lie in that bucket. For eg. if there are 3 replicas with 4 cores each, i.e., total 12 cores, then this configuration could belong to a bucket with:
+```
+minimum CPU: 3
+maximum CPU: 6
+minimum memory: 2.5G
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
```

Comment:
>You want hysteresis

Discussion about hysteresis continued [here](https://github.com/gardener/hvpa-controller/pull/71#discussion_r430107943).

Reply:
> we prefer vertical scaling within the bucket first

>...overlapping buckets within the user-provided buckets. Is that what is suggested here?

Yes, this is indeed the case. Let's consider the example shown in the doc:
```
minimum CPU: 3
maximum CPU: 6
minimum memory: 2.5G
maximum memory: 7G
minimum replicas: 2
maximum replicas: 5
```
Current: replicas: 2, cpu: 3.2, memory: 2.7G -  total CPU = 6.4, total memory: 5.4G
The upper bound for 2 replicas in this “linearly” divided bucket would be:
Cpu: `3 + (6-3)/(5-2+1) = 3.75`
Memory: `2.5 + (7-2.5)/(5-2+1) = 3.6G`

This calculation is explained using the [chart](https://github.com/gardener/hvpa-controller/blob/e42e7c9430c5dd8502f93e3ed14a2b37f58c20c9/Proposals/images/CPUStory1.png) in this doc.

So unless the aggregate recommendation go above `2 * 3.75 = 7.5 CPU` or `2 * 3.6 = 7.2G memory`, HVPA controller would scale the application only vertically keeping the replica count at 2.

> I believe, that would bring us to too many replicas too soon.

It would, but only if the range of vertical scale is too less. To see how it would work for kube-apiserver according to what we discussed, please refer the section [kube-apiserver scaling example
](https://github.com/gardener/hvpa-controller/blob/e42e7c9430c5dd8502f93e3ed14a2b37f58c20c9/Proposals/20200511-HVPA-design.md#kube-apiserver-scaling-example)
Here, `minValue` and `maxValue` for replicas are equal in each bucket. So the bucket would not be "linearly divided", and the upper cap at each replica count is the actual `maxValue` of the related bucket.

Reply:
>Otherwise things get even more complicated, because you then need to have overlapping buckets within the user-provided buckets. Is that what is suggested here?

Yes, this is what we had discussed. We called them linearly/exponentially/logarithmically divided buckets.
@amshuman-kr